lib/common_test/src/ct.erl

```diff
@@ -89,34 +92,135 @@
+           TestDirs :: [string()] | string(),
+           Suites :: [string()] | [atom()] | string() | atom(),
+           Cases :: [atom()] | atom(),
+           Groups :: GroupNameOrPath | [GroupNameOrPath],
+           GroupNameOrPath :: [atom()] | atom() | all,
+           TestSpecs :: [string()] | string(),
+           Label :: string() | atom(),
+           CfgFiles :: [string()] | string(),
+           UserConfig :: [{CallbackMod, CfgStrings}] | {CallbackMod, CfgStrings},
+           CallbackMod :: atom(),
+           CfgStrings :: [string()] | string(),
+           LogDir :: string(),
+           Conns :: all | [atom()],
+           CSSFile :: string(),
+           CoverSpecFile :: string(),
+           StepOpts :: [StepOpt] | [],
+           StepOpt :: config | keep_inactive,
+           EventHandlers :: EH | [EH],
+           EH :: atom() | {atom(), InitArgs} | {[atom()], InitArgs},
+           InitArgs :: [term()],
+           InclDirs :: [string()] | string(),
+           CreatePrivDir :: auto_per_run | auto_per_tc | manual_per_tc,
+           M :: integer(),
+           N :: integer(),
+           DurTime :: HHMMSS,
+           HHMMSS :: string(),
+           StopTime :: YYMoMoDDHHMMSS | HHMMSS,
+           YYMoMoDDHHMMSS :: string(),
+           ForceStop :: skip_rest | Bool,
+           DecryptKeyOrFile :: {key, DecryptKey} | {file, DecryptFile},
+           DecryptKey :: string(),
+           DecryptFile :: string(),
+           LogOpts :: [LogOpt],
+           LogOpt :: no_nl | no_src,
+           VLevels :: VLevel | [{Category, VLevel}],
+           VLevel :: integer(),
+           Category :: atom(),
+           KeepSpec :: all | pos_integer(),
+           CTHs :: [CTHModule | {CTHModule, CTHInitArgs}],
+           CTHModule :: atom(),
+           CTHInitArgs :: term(),
+           Result :: {Ok, Failed, {UserSkipped, AutoSkipped}} | TestRunnerPid | {error, Reason},
+           Ok :: integer(),
+           Failed :: integer(),
+           UserSkipped :: integer(),
+           AutoSkipped :: integer(),
+           TestRunnerPid :: pid(),
+           Reason :: term().
 run_test(Opts) ->
     ct_run:run_test(Opts).
 
+-spec run_testspec(TestSpec) -> Result
+      when TestSpec :: [term()],
+           Result :: {Ok, Failed, {UserSkipped, AutoSkipped}} | {error, Reason},
+           Ok :: integer(),
+           Failed :: integer(),
+           UserSkipped :: integer(),
+           AutoSkipped :: integer(),
+           Reason :: term().
 run_testspec(TestSpec) ->
     ct_run:run_testspec(TestSpec).
     
+-spec step(TestDir, Suite, Case) -> Result
+      when TestDir :: string(),
+           Suite :: atom(),
+           Case :: atom(),
+           Result :: term().
 step(TestDir,Suite,Case) ->
     ct_run:step(TestDir,Suite,Case).
 
+-spec step(TestDir, Suite, Case, Opts) -> Result
+      when TestDir :: string(),
+           Suite :: atom(),
+           Case :: atom(),
+           Opts :: [Opt] | [],
```

Comment:
the empty list is already contained in a `[Opt]`, so `Opts :: [Opt],`

Reply:
d02030ecd2